### PR TITLE
Feature/optimize game memory

### PR DESCRIPTION
**Pull request by misterokaygo**

### CODE REVIEW

File: Helpers/GameMemory.cs [hunk 1]
```diff
@@ -58,101 +51,59 @@ public static GameData GetGameData()
                     WindowsExternal.OpenProcess((uint)WindowsExternal.ProcessAccessFlags.VirtualMemoryRead, false, gameProcess.Id);
                 IntPtr processAddress = gameProcess.MainModule.BaseAddress;
 
-                if (PtrPlayerUnit == IntPtr.Zero)
+                if (Equals(PlayerUnit, default(UnitAny)))
                 {
-                    IntPtr pUnitTable = IntPtr.Add(processAddress, Offsets.UnitTable);
-                    for (var i = 0; i < 128; i++)
+                    var unitHashTable = Read<UnitHashTable>(processHandle, IntPtr.Add(processAddress, Offsets.UnitHashTable));
+                    for (var i = 0; i < unitHashTable.UnitTable.Length; i++)
```
Comment (AudriusButkevicius) [inline Helpers/GameMemory.cs:57]: foreach?

File: Helpers/GameMemory.cs [hunk 2]
```diff
@@ -58,101 +51,59 @@ public static GameData GetGameData()
...
-                IntPtr pRoom2 = IntPtr.Add(aRoom1, 0x18);
-
-                WindowsExternal.ReadProcessMemory(processHandle, pRoom2, addressBuffer, addressBuffer.Length, out _);
-                IntPtr aRoom2 = (IntPtr)BitConverter.ToInt64(addressBuffer, 0);
-                IntPtr pLevel = IntPtr.Add(aRoom2, 0x90);
-
-                WindowsExternal.ReadProcessMemory(processHandle, pLevel, addressBuffer, addressBuffer.Length, out _);
-                IntPtr aLevel = (IntPtr)BitConverter.ToInt64(addressBuffer, 0);
-                IntPtr aLevelId = IntPtr.Add(aLevel, 0x1F8);
+                var playerName = Encoding.ASCII.GetString(Read<byte>(processHandle, PlayerUnit.UnitData, 16)).TrimEnd((char)0);
```
Comment (AudriusButkevicius) [inline Helpers/GameMemory.cs:81]: Don't think you need the trim end. GetString should handle it.

File: Helpers/GameMemory.cs [hunk 3]
```diff
@@ -58,101 +51,59 @@ public static GameData GetGameData()
...
+                var act = Read<Act>(processHandle, (IntPtr)PlayerUnit.pAct);
```
Comment (AudriusButkevicius) [inline Helpers/GameMemory.cs:82]: 🤮 🤮 🤮  imagine just reading IntPtrs straight away?

File: MapAssist.csproj [hunk 4]
```diff
@@ -9,7 +9,8 @@
     <ProjectGuid>{FB3AC437-B7BB-42FF-A2A2-F10551F95F6C}</ProjectGuid>
     <OutputType>WinExe</OutputType>
     <RootNamespace>MapAssist</RootNamespace>
-    <CustomAssemblyName></CustomAssemblyName>
+    <CustomAssemblyName>
+    </CustomAssemblyName>
```
Comment (AudriusButkevicius) [inline MapAssist.csproj:13]: Will break rider.

File: Helpers/GameMemory.cs (hunk 1 above)
Comment (OneXDeveloper) [inline Helpers/GameMemory.cs:57]: Player is UnitType 0 so we should be able to skip any types that aren't 0.

File: Structs/Level.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ **/
+
+using System.Runtime.InteropServices;
+
+namespace MapAssist.Structs
+{
+    [StructLayout(LayoutKind.Explicit)]
+    unsafe public struct Level
+    {
+        [FieldOffset(0x1F8)] public uint LevelId;
```
Comment (AudriusButkevicius) [inline Structs/Level.cs:27]: Could use Area enum.

File: Structs/Room.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ **/
+
+using System.Runtime.InteropServices;
+
+namespace MapAssist.Structs
+{
+    [StructLayout(LayoutKind.Explicit)]
+    unsafe public struct Room
+    {
+        [FieldOffset(0)] public Room** pRoomsNear;
```
Comment (AudriusButkevicius) [inline Structs/Room.cs:27]: 0x00

File: Helpers/GameMemory.cs (hunk 1 above)
Comment (AudriusButkevicius) [reply Helpers/GameMemory.cs:57]: In theory there shouldn't be any non-players in this table. In theory.

**Review by AudriusButkevicius — APPROVED** [review]
lgtm

**Pull request merged**